File: ASD/strutture_dati/pila.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>


//Algoritmo FIFO

typedef struct node{
    int e;
    struct node *next;
}node;

typedef struct pila{   
    struct node *start;
    int len;

}pila;


node* init_node(int value){
    node *nd = malloc(sizeof(node));
    
    nd->e = value;
    nd->next = NULL;

    return nd;
}

pila* init_pila(){
    pila *P = malloc(sizeof(pila));

    P->start = NULL;
    P->len = 0;

    return P;
}

void printer(pila *P){
    printf("[");
    node *nd = P->start;
    while (nd != NULL) {
        printf("%d", nd->e);
        if (nd->next != NULL) {
            printf(", ");
        }
        nd = nd->next;
    }
    printf("]\n");
}

bool isEmpty(pila *P){
    return P->start == NULL;
}


void push(pila *P, int e){
    node *nd = init_node(e);
    
    nd->next = P->start;
    P->start = nd;
    P->len++;

}

int pop(pila * P){
    node *nd = P->start;

    if(nd == NULL){
        printf("Bru la pila è vuota: ");
        return -1;
    }else{
        int e = P->start->e;
        P->start = P->start->next;
        free(nd);
        P->len--;
        return e;
    }

}

int top(pila *P){
    return P->start->e;
}

int main(){
    pila *P = init_pila();

    push(P, 10);
    push(P, 20);
    push(P, 30);

    printf("Pila dopo gli accodamenti: ");
    printer(P);

    printf("%d\n",pop(P));
    printf("%d\n",pop(P));
    printf("%d\n",pop(P));
    printf("%d\n",pop(P));

    printf("Pila dopo le cancellazioni: ");
    printer(P);

    
    return 0;
}

File: ASD/strutture_dati/pila.py
def isEmpty(P,len_P):
    return len_P == 0

def push(P, len_P, elem):
    if len_P < len(P):
        P[len_P] = elem        
    else:
        P.append(elem)
    len_P += 1    
    return len_P

def pop(P, len_P):
    if isEmpty(P,len_P):
        return None, len_P

    e = P[len_P - 1]
    P[len_P - 1] = None
    len_P -= 1

    if len_P < len(P)/2:
            new_P  = []
            for i in range(len_P):
                new_P.append(P[i])
            P = new_P

    return P, e , len_P

def top(P, len_P):
    if isEmpty(P,len_P):
        return None
    
    return P[len_P - 1]

P  = []
len_P = len(P)
print(f"La lista è vuota -> {isEmpty(P, len_P)}")
for i in range(10):
    len_P = push(P, len_P, i)

print(f"Lista dopo gli inserimento -> {P}")
print(f"La lista è vuota -> {isEmpty(P, len_P)}")

for i in range(7):
    P, e, len_P = pop(P,len_P)
    print(f"Ho eliminato l'elemento -> {e}") 

print(f"Lista dopo le cancellazione -> {P}")

push(P, len_P, 10)
print(f"Lista dopo l'ultimo inserimento -> {P}")

print(f"L'elemento in cima è -> {top(P,len_P)}")
